src/operator/contrib/hawkes_ll-inl.h

```diff
@@ -0,0 +1,493 @@
+                  betag_[m] - alpha[m] * (
+                    phig_buffer[i] * (1 - expf(-beta[m] * d))
+                    + phi_buffer[i] * d * expf(-beta[m] * d)
+                  )
+    )
+
+    // // correct the gradients with respect to output gradients
+    KERNEL_ASSIGN(mug_[m], req, out_grad[j] * mug_[m])
+    KERNEL_ASSIGN(alphag_[m], req, out_grad[j] * alphag_[m])
+    KERNEL_ASSIGN(betag_[m], req, out_grad[j] * betag_[m])
+  }
+};
+
+template<typename xpu>
+void HawkesLLBackward(const nnvm::NodeAttrs& attrs,
+                         const OpContext& ctx,
+                         const std::vector<TBlob>& inputs,
+                         const std::vector<OpReqType>& req,
+                         const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 10U);
+  CHECK_EQ(outputs.size(), 8U);
+  CHECK_EQ(req.size(), 8U);
+
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+
+  int K = inputs[hawkesll::kGradMu].shape_[1];  // mu data
+  int N = inputs[hawkesll::kGradIATimes].shape_[0];
+  int T = inputs[hawkesll::kGradIATimes].shape_[1];
+
+  CHECK_EQ(inputs[hawkesll::kOutGradLL].shape_[0], N);  // grad of out 0 (LL)
+  CHECK_EQ(inputs[hawkesll::kOutGradStates].shape_[0], N);  // grad out 1-states
+  CHECK_EQ(inputs[hawkesll::kOutGradStates].shape_[1], K);
+
+  const TBlob& out_grad = inputs[hawkesll::kOutGradLL];
+
+  using namespace mshadow;
+  using namespace mxnet_op;
+  MSHADOW_TYPE_SWITCH(out_grad.type_flag_, DType, {
+    // allocate gradient buffers
+    Tensor<xpu, 2, DType> bfr =
+        ctx.requested[hawkesll::kTempSpace]
+           .get_space_typed<xpu, 2, DType>(Shape2(6*N, K), s);
+
+    Tensor<xpu, 2, DType> alpha_gbfr =
+        Tensor<xpu, 2, DType>(&bfr.dptr_[N*K], Shape2(N, K), s);
+    Tensor<xpu, 2, DType> beta_gbfr =
+        Tensor<xpu, 2, DType>(&bfr.dptr_[2*N*K], Shape2(N, K), s);
+    Tensor<xpu, 2, DType> last_buffer =
+        Tensor<xpu, 2, DType>(&bfr.dptr_[3*N*K], Shape2(N, K), s);
+    Tensor<xpu, 2, DType> phig_buffer =
+        Tensor<xpu, 2, DType>(&bfr.dptr_[4*N*K], Shape2(N, K), s);
+    Tensor<xpu, 2, DType> phi_buffer =
+        Tensor<xpu, 2, DType>(&bfr.dptr_[5*N*K], Shape2(N, K), s);
+
+    alpha_gbfr = DType(0.0);
+    beta_gbfr = DType(0.0);
+    last_buffer = DType(0.0);
+    phig_buffer = DType(0.0);
+
+    mshadow::Copy(phi_buffer,
+                  inputs[hawkesll::kGradState]
+                    .get_with_shape<xpu, 2, DType>(Shape2(N, K), s),
+                  s);
+
+    // get the gradient to be output
+    Tensor<xpu, 2, DType> in_grad_mu =
+        outputs[hawkesll::kMu].get_with_shape<xpu, 2, DType>(Shape2(N, K), s);
+    Tensor<xpu, 1, DType> in_grad_alpha =
+        outputs[hawkesll::kAlpha].get_with_shape<xpu, 1, DType>(Shape1(K), s);
+    Tensor<xpu, 1, DType> in_grad_beta =
+        outputs[hawkesll::kBeta].get_with_shape<xpu, 1, DType>(Shape1(K), s);
+
+    in_grad_mu = DType(0.0);
+
+    MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
```

Comment:
`req` is a list of grad_reqs that indicates whether it is required to compute the gradient r.s.t. input[0], input[1], ... input[7] in the forward function. Is it possible that they have different values (req[0] = 'write', req[1] = 'add_to', req[2] = 'null'). 

Specifically, `MXNET_ASSIGN_REQ_SWITCH ` returns immediately if req = 'null': https://github.com/apache/incubator-mxnet/blob/2fd4720342f87be80f280176ffbecb8f37e49b73/src/operator/mxnet_op.h#L92 

It looks like the kernel assume the same value for all reqs. If so, can you add a CHECK_EQ statement for all these reqs? Thanks.

Reply:
Thank you so much! That saves me from some trouble :bowtie:.

The operator computes gradients for the first three inputs only. The current version checks for each of the three separately, and makes sure the rest are null with `CHECK_EQ`.